dnadoh/cache.py

```diff
@@ -0,0 +1,118 @@
+3.  Make `FileCache` able to download both remote files and
+    local files.
+4.  Check that files whose names are in the cache are actually
+    stored locally rather than trusting the cache.
+5.  Rebuild the cache from the files that we already have on
+    disk, and explain why this is both useful and dangerous.
+6.  Explain why we need to be able to clear the cache. (Hint:
+    think about testing.)
+7.  Discuss why `get` and `clear` are `classmethod` rather than
+    `staticmethod`.  # because depends on the cache path?
+"""
+
+import filecmp
+import requests
+import shutil
+import validators
+
+from pathlib import Path
+from urllib.parse import urljoin
+
+
+CONFIG = {
+    "cache": Path(__file__).parent.parent.joinpath("cache"),  # prints DNA-doh rootdir + cache
+    "remote": Path(__file__).parent.parent.joinpath("remote"),
+}
+
+
+class FileCache:
+    """Cache large remote files."""
+
+    def __init__(self, cache_dir=None, remote_url=None):
+        if cache_dir:
+            self.cache_dir = cache_dir
+        else:
+            self.cache_dir = CONFIG["cache"]
+        
+        if remote_url:
+            self.remote_url = remote_url
+        else:
+            self.remote_url = CONFIG["remote"]
+        
+        self.files = {}
+        
+    def get(self, filename):
+        """Return path to cached copy of file, getting as needed."""
+        if filename not in self.files:
+            # check if valid URL, which implies it can be downloaded
+            if validators.url(filename) is True:
+                # extact filename from URL
+                extracted_filename = filename.split('/')[-1]
+                cache_path = Path(self.cache_dir, extracted_filename)
+                if self._download_file(filename, cache_path):
+                    self.files[filename] = cache_path
+                else:
+                    # download failed
+                    print(f"{filename} could not be downloaded.")
+            else:
+                cache_path = Path(self.cache_dir, filename)
+                self._get_file(filename, cache_path)
+                self.files[filename] = cache_path
+        else:
+            # 4 - do a further check that the file is actually the same
+            if filecmp.cmp(Path(self.cache_dir, filename), Path(self.remote_url, filename)) is True:  # this doesn't work for remote URLs
+                return self.files[filename]
+            else:
+                self._get_file(filename, cache_path)
+                self.files[filename] = cache_path
+
+    def clear(self):
+        """Clear the cache, deleting local files."""
+        self.files.clear()
+        for filename in self.cache_dir.iterdir():
+            filename.unlink()
+
+    def _get_file(self, remote_path, local_path):
+        """Simulate getting a remote file."""
+        remote_path = Path(self.remote_url, remote_path)
+        # print(remote_path)
+        assert remote_path.exists(), f"{remote_path} does not exist"
```

Comment:
production code could/should raise some other kind of exception rather than `AssertionError` (I use that when I'm lazy, but it should only be used for internal programming errors, and missing files are almost always someone's actual fault)

Reply:
I used some pretty basic Exceptions - do you think they're good enough/make sense in this context? Or do we need to make our own exceptions (that'd seem like overkill for this IMO)